lib/Conversion/TorchToLinalg/Pooling.cpp

```diff
@@ -36,94 +137,180 @@ class ConvertAtenMaxPool2dOp : public OpConversionPattern<AtenMaxPool2dOp> {
-        rewriter, loc, W, paddingIntValues[1], dilationIntValues[1],
-        kernelSizeIntValues[1], strideIntValues[1]);
-
-    // Initialize output tensor with smallest floating point value
-    Value outTensor = rewriter.create<linalg::InitTensorOp>(
-        loc, ValueRange{N, C, Hout, Wout}, elementType);
-    auto initialAttr = rewriter.getFloatAttr(
-        elementType,
-        APFloat::getSmallest(
-            elementType.cast<mlir::FloatType>().getFloatSemantics(),
-            /*Negative*/ true));
-    Value initValue = rewriter.create<arith::ConstantOp>(loc, initialAttr);
-    Value outTensorInitialized =
-        rewriter.create<linalg::FillOp>(loc, initValue, outTensor).getResult(0);
-
-    auto stridesAttr = rewriter.getI64VectorAttr(strideInts);
-    auto dilationAttr = rewriter.getI64VectorAttr(dilationInts);
+    SmallVector<Value> resultShape(getTensorSizes(rewriter, loc, maxPool2d));
+    Value indicesTensor = createZeroInitTensor(
+        rewriter, loc, resultShape, indicesRankedTensorType.getElementType());
+
+    SmallVector<Value> kernelSize =
+        getAsConstantIndexValues(rewriter, loc, kernelSizeInts);
+    SmallVector<Value> padding =
+        getAsConstantIndexValues(rewriter, loc, paddingInts);
+    SmallVector<Value> dilation =
+        getAsConstantIndexValues(rewriter, loc, dilationInts);
+    SmallVector<Value> stride =
+        getAsConstantIndexValues(rewriter, loc, strideInts);
+
     Value windowTensor = rewriter.create<linalg::InitTensorOp>(
-        loc, getAsConstantIndexValues(rewriter, loc, kernelSizeInts),
-        elementType);
-
-    Value maxPool2d = rewriter
-                          .create<linalg::PoolingNchwMaxOp>(
-                              loc, outTensorInitialized.getType(),
-                              ValueRange{paddedInput, windowTensor},
-                              outTensorInitialized, stridesAttr, dilationAttr)
-                          .getResult(0);
-    Type newResultType = getTypeConverter()->convertType(op.getType());
-    rewriter.replaceOpWithNewOp<tensor::CastOp>(op, newResultType, maxPool2d);
+        loc, kernelSize, indicesRankedTensorType.getElementType());
+
+    SmallVector<AffineExpr> inputExprs, outputExprs, kernelExprs;
+    for (unsigned i = 0; i < 4; i++) {
+      inputExprs.push_back(rewriter.getAffineDimExpr(i));
+      outputExprs.push_back(rewriter.getAffineDimExpr(i));
+    }
+    kernelExprs.push_back(rewriter.getAffineDimExpr(4));
+    kernelExprs.push_back(rewriter.getAffineDimExpr(5));
+
+    // Here we have six dimensions, each corresponding to N, C, Hout, Wout, kH,
+    // and kW, respectively, as described in the algorithm above.
+    SmallVector<AffineMap> indexingMaps = {
+        AffineMap::get(6, 0, inputExprs, op.getContext()),
+        AffineMap::get(6, 0, kernelExprs, op.getContext()),
+        AffineMap::get(6, 0, outputExprs, op.getContext())};
+    SmallVector<StringRef> iteratorTypes(6, getParallelIteratorTypeName());
```

Comment:
If a max value appeared more than once in the window would the smallest index would be returned. I think the iteration along kH and kW need to be sequential because of such cases.